test/mock/adapter/adapter.go

```diff
@@ -1127,15 +1119,192 @@ func (v *adapterApp) openid4vcInitiatePreAuthorizedIssuance(w http.ResponseWrite
 		return
 	}
 
+	var credentialsToSave map[string]json.RawMessage
+	err = json.Unmarshal([]byte(credentials), &credentialsToSave)
+	if err != nil {
+		handleError(w, http.StatusInternalServerError,
+			fmt.Sprintf("failed to parse credentials : %s", err))
+
+		return
+	}
+
+	for ct, credential := range credentialsToSave {
+		err = v.store.Put(getCredStoreKeyPrefix(key, ct), credential)
+		if err != nil {
+			handleError(w, http.StatusInternalServerError,
+				fmt.Sprintf("failed to server configuration : %s", err))
+
+			return
+		}
+	}
+
 	initiateUrl := "openid-initiate-issuance://?issuer=" + url.QueryEscape(issuer) + "&credential_type=" +
-		url.QueryEscape(credType) + "&pre-authorized_code=" + url.QueryEscape(uuid.NewString()) + "&user_pin_required=true"
+		url.QueryEscape(credType) + "&pre-authorized_code=" + url.QueryEscape(preAuthCode) + "&user_pin_required=true"
 
 	err = t.Execute(w, openid4ciDemo{
 		InitiateUrl: initiateUrl,
 		Pin:         pinNumber,
 	})
 }
 
+func (v *adapterApp) openid4vcIssuerTokenEndpoint(w http.ResponseWriter, r *http.Request) {
+	setOIDCResponseHeaders(w)
+
+	code := r.FormValue("pre-authorized_code")
+	userPin := r.FormValue("user_pin")
+	grantType := r.FormValue("grant_type")
+
+	if grantType != "urn:ietf:params:oauth:grant-type:pre-authorized_code" {
+		sendOIDCErrorResponse(w, "unsupported grant type", http.StatusBadRequest)
+		return
+	}
+
+	authRqstBytes, err := v.store.Get(getPreAuthCodeKeyPrefix(code))
+	if err != nil {
+		sendOIDCErrorResponse(w, "invalid request", http.StatusBadRequest)
+		handleError(w, http.StatusInternalServerError,
+			fmt.Sprintf("failed to fetch auth state : %s", err))
+		return
+	}
+
+	var authRequest map[string]string
+	err = json.Unmarshal(authRqstBytes, &authRequest)
+	if err != nil {
+		sendOIDCErrorResponse(w, "failed to read request", http.StatusInternalServerError)
+		return
+	}
+
+	if userPin != authRequest["pin"] {
```

Comment:
Actually `pin` is optional

Reply:
Yes, but if the issuer sets `user_pin_required = true` in the initiate issuance response, then `pin` is required. We could make this more flexible by customizing the issuer, if needed